### ISSUE TITLE
Regularizing and Optimizing LSTM Language Models

### ISSUE THREAD

**Issue by amyxie361** [amyxie361/Blog#1]
# Regularizing and Optimizing LSTM Language Models

*Authors*: Stephen Merity 1 Nitish Shirish Keskar 1 Richard Socher 1
*Arxiv*: https://arxiv.org/pdf/1708.02182v1.pdf
Task: work level language model, WikiText2, perplexity.

# Survey part
### Dropout
	- naive dropout(sampling a new binary mask at each timestep ) - disrupts RNN’s ability to retain long term dependency
	- same dropout mask across multiple time steps
	- limit updates tot eh RNN’s hidden state - drop updates(input gates) - zone-out

### Restrictions on the recurrent matrices
	- restricting the capacity of the matrix
	- element-wise interaction

### Others
	- batch normalization
	- recurrent batch normalization
	- layer normalization
 - Introduce additional training parameters 

# Propose: 

## DropConnect
Mask on the hidden-to-hidden recurrent weights

BPTT: randomized-length backpropagation through time
Embedding dropout
Activation regularization
Temporal activation regularization

？为什么用LSTM不用GRU
![image](https://user-images.githubusercontent.com/26669286/71130196-5cf12880-21bf-11ea-90a1-bc0d28c10b1b.png)
![image](https://user-images.githubusercontent.com/26669286/71130220-667a9080-21bf-11ea-88f0-44ec73425186.png)

## Optimizer - NT-ASGD
SGD works well	
ASGD
NT-SGD
![image](https://user-images.githubusercontent.com/26669286/71130169-52369380-21bf-11ea-816f-3f8eef7a1f0b.png)

# Experiment
尝试了不同的regularization方法
1. Variable-length backpropagation sequences:
	固定截断seq长度的话会导致训练数据里面的固定位置很少收到后续token的信号
	改进：window的选取服从高斯分布
2. Variational dropout
	不同gate用不同的dropout mask， 同一个train step的不同time step间用相同的mask
3. Embedding dropout
	每个train step随机扔掉一些词
4. Weight tying
	输入word embedding 和输出softmax前的weight共享参数
5. Independent embedding size and hidden size
	压缩LSTM hidden size减参数量
6. AR / TAR
	AR: mask normalization里的参数
	TAR: 给变化大的hidden加惩罚
7. Neural cache model
	存之前的cell state 做average

![image](https://user-images.githubusercontent.com/26669286/71130254-77c39d00-21bf-11ea-953f-eaf961cfb305.png)

# Comment
实验结果现实比之前的模型都好
有好些地方不太懂他为什么选了这个方法做实验而不是其他的
比较的时候为什么没有给training过程中的效果展示，只是展示了最后的acc

# 值得看的其他reference

- [ ] Dropout: a simple way to prevent neural networks from overfitting. Srivastava
- [ ] Batch normalization: Accelerating deep network training by reducing internal covariate
shift. Ioffe
- [ ] A theoretically grounded application of dropout in recurrent neural networks. Gal
- [ ] Zoneout: Regularizing RNNss by randomly preserving hidden activations. Krueger
- [ ]  This can be done either through restricting the capacity of the matrix (Arjovsky et al.,
2016; Wisdom et al., 2016; Jing et al., 2016) or through element-wise interactions (Balduzzi & Ghifary, 2016; Bradbury et al., 2016; Seo et al., 2016).
- [ ] recurrent batch normalization (Cooijmans et al., 2016)
- [ ] layer normalization (Ba et al., 2016)
- [ ] backpropagation through time (BPTT),
- [ ] Optimization methods for large-scale machine learning. Bottou
- [ ] Gradient descent converges to minimizers: The case of non-isolated critical points. Panageas, I. and Piliouras, G. 
- [ ] Train faster, generalize better: Stability of stochastic gradient descent. Hardt, M., Recht, B., and Singer, Y. 
- [ ] Improving neural language models with a continuous cache. Grave, E., Joulin, A., and Usunier, N.
